src/charm.py

```diff
@@ -0,0 +1,595 @@
+        """Handle the installation of charm.
+
+        Args:
+            event: Event of installing charm.
+        """
+        self.unit.status = MaintenanceStatus("Installing packages")
+
+        try:
+            # The `_start_services`, `_install_deps` includes retry.
+            self._install_deps()
+            self._start_services()
+        except SubprocessError as err:
+            logger.exception(err)
+            # The charm cannot proceed without dependencies.
+            self.unit.status = BlockedStatus("Failed to install dependencies")
+            return
+
+        runner_manager = self._get_runner_manager()
+        if runner_manager:
+            self.unit.status = MaintenanceStatus("Downloading runner binary")
+            try:
+                runner_info = runner_manager.get_latest_runner_bin_url()
+                logger.info(
+                    "Downloading %s from: %s", runner_info.filename, runner_info.download_url
+                )
+                self._stored.runner_bin_url = runner_info.download_url
+                runner_manager.update_runner_bin(runner_info)
+            # Safe guard against transient unexpected error.
+            except Exception as err:  # pylint: disable=broad-exception-caught
+                logger.exception("Failed to update runner binary")
+                # Failure to download runner binary is a transient error.
+                # The charm automatically update runner binary on a schedule.
+                self.unit.status = MaintenanceStatus(f"Failed to update runner binary: {err}")
+                return
+            self.unit.status = MaintenanceStatus("Starting runners")
+            try:
+                self._reconcile_runners(runner_manager)
+                self.unit.status = ActiveStatus()
+            except RunnerError as err:
+                logger.exception("Failed to start runners")
+                self.unit.status = MaintenanceStatus(f"Failed to start runners: {err}")
+        else:
+            self.unit.status = BlockedStatus("Missing token or org/repo path config")
+
+    @catch_unexpected_charm_errors
+    def _on_upgrade_charm(self, _event: UpgradeCharmEvent) -> None:
+        """Handle the update of charm.
+
+        Args:
+            event: Event of charm upgrade.
+        """
+        logger.info("Reinstalling dependencies...")
+        self._install_deps()
+        self._start_services()
+
+        logger.info("Flushing the runners...")
+        runner_manager = self._get_runner_manager()
+        if not runner_manager:
+            return
+
+        runner_manager.flush()
+        self._reconcile_runners(runner_manager)
+
+    @catch_unexpected_charm_errors
+    def _on_config_changed(self, _event: ConfigChangedEvent) -> None:
+        """Handle the configuration change.
+
+        Args:
+            event: Event of configuration change.
+        """
+        try:
+            self._event_timer.ensure_event_timer(
```

Comment:
I guess this is a way to periodically trigger custom events? Not a bad pattern until perhaps pebble notixes if we can find a better source for these events than cron

Reply:
@benhoyt any comment?

Reply:
Yeah, this seems like a nice workaround!